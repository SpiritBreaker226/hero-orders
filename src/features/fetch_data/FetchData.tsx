import React, { FunctionComponent, useEffect } from 'react'
import { useDispatch, useSelector } from 'react-redux'

import axios from 'axios'

import {
  updateLoading,
  updateErrorMessageFromServer,
  selectFetch,
} from './fetchDataSlice'

const FetchData: FunctionComponent = () => {
  const dispatch = useDispatch()
  const state = useSelector(selectFetch)

  useEffect(() => {
    const fetchData = async () => {
      try {
        dispatch(updateLoading(true))

        await axios.get(state.urlEndpoint)

        dispatch(updateLoading(false))
      } catch (error) {
        dispatch(
          updateErrorMessageFromServer({
            isLoading: false,
            errorMessage: error.message,
          })
        )
      }
    }

    fetchData()
  }, [dispatch, state.urlEndpoint])

  if (state.isLoading) return <div className="message">Loading...</div>
  if (state.errorMessage) {
    return <div className="message">{state.errorMessage}</div>
  }

  return null
}

export default FetchData
