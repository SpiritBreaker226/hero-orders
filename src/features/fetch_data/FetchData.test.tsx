import React from 'react'

import { waitFor, screen, render } from '@testing-library/react'

import axios from 'axios'

import { Provider } from 'react-redux'
import { createStore } from '@reduxjs/toolkit'

import FetchData from './FetchData'
import reducer, { initialState } from './fetchDataSlice'

jest.mock('axios')
const mockedAxios = axios as jest.Mocked<typeof axios>

describe('FetchData', () => {
  describe('fetch Order data', () => {
    it('renders loading screen', async () => {
      const mockStore = createStore(reducer, {
        fetch: {
          ...initialState,
          isLoading: true,
        },
      })

      render(
        <Provider store={mockStore}>
          <FetchData />
        </Provider>
      )

      expect(screen.getByText('Loading', { exact: false })).toBeInTheDocument()
      expect(screen.getByText('Loading', { exact: false })).toMatchSnapshot()

      await waitFor(() => {})
    })

    describe('on error from the server', () => {
      it('should render an error message', async () => {
        mockedAxios.get.mockRejectedValue(new Error('fake error message'))

        const mockStore = createStore(reducer, {
          fetch: {
            ...initialState,
            errorMessage: 'fake error message',
            isLoading: false,
          },
        })

        render(
          <Provider store={mockStore}>
            <FetchData />
          </Provider>
        )

        await waitFor(() => screen.getByText('error', { exact: false }))

        const error = screen.getByText('fake error message', { exact: false })

        expect(error).toBeInTheDocument()
        expect(error).toMatchSnapshot()
      })
    })
  })
})
