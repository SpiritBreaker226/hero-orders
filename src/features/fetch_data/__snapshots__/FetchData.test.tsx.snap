// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FetchData fetch Order data on error from the server should render an error message 1`] = `
<div
  class="message"
>
  fake error message
</div>
`;

exports[`FetchData fetch Order data renders loading screen 1`] = `
<div
  class="message"
>
  Loading...
</div>
`;
